dango/dex/src/core/liquidity_pool.rs

```diff
@@ -162,27 +176,67 @@ impl PassiveLiquidityPool for PairParams {
+        // without paying any fee.
+        //
+        // We determine the fee rate based oracle price as follows:
+        //
+        // - Suppose the pool's reserve is `A` dollars of the 1st asset and `B`
+        //   dollars of the 2nd asset.
+        // - Suppose a user deposits `a` dollars of 1st asset and `b` dollars of
+        //   the 2nd asset.
+        // - Note that `A`, `B`, `a`, and `b` here are values in USD, not the
+        //   unit amounts.
+        // - Without losing generality, assume a / A > b / B. In this case, the
+        //   user is over-supplying the 1st asset and under-supplying the 2nd
+        //   asset. To make the deposit symmetrical, the user needs to swap some
+        //   1st asset into some 2nd asset.
+        // - Suppose user swap `x` dollars of the 1st asset into the 2nd asset.
+        //   Also suppose our pool does the swap at exactly the oracle price
+        //   without slippage. This assumption obviously isn't true, but is good
+        //   enough for the purpose of preventing the aforementioned attack.
+        // - We must solve:
+        //   (a - x) / (A + x) = (b + x) / (B - x)
+        //   The solution is:
+        //   x = (a * B - A * b) / (a + A + b + B)
+        // - We charge a fee assuming this swap is to be carried out. The USD
+        //   value of the fee is:
+        //   x * swap_fee_rate
+        // - To charge the fee, we mint slightly less LP tokens corresponding to
+        //   the ratio:
+        //   x * swap_fee_rate / (a + b)
+        //
+        // Related: Curve V2 also applies a fee for asymmetrical deposits:
         // https://github.com/curvefi/twocrypto-ng/blob/main/contracts/main/Twocrypto.vy#L1146-L1168
-        let (a, b, reserve_a, reserve_b) = (
-            *deposit.first().amount,
-            *deposit.second().amount,
-            *reserve.first().amount,
-            *reserve.second().amount,
-        );
+        // However, their math appears to be only suitable for the Curve V2 curve.
+        // Our oracle approach is more generalizable to different pool types.
+        let fee_rate = {
+            let price = oracle_querier.query_price(reserve.first().denom, None)?;
+            let a = price.value_of_unit_amount(*deposit.first().amount)?;
+            let reserve_a = price.value_of_unit_amount(*reserve.first().amount)?;
+
+            let price = oracle_querier.query_price(reserve.second().denom, None)?;
+            let b = price.value_of_unit_amount(*deposit.second().amount)?;
+            let reserve_b = price.value_of_unit_amount(*reserve.second().amount)?;
+
+            let deposit_value = a.checked_add(b)?;
+            let reserve_value = reserve_a.checked_add(reserve_b)?;
+
+            abs_diff(a.checked_mul(reserve_b)?, b.checked_mul(reserve_a)?)
+                .checked_div(deposit_value.checked_add(reserve_value)?)?
+                .checked_mul(*self.swap_fee_rate)?
+                .checked_div(deposit_value)?
+        };
```

Comment:
In terms of code styling, I like using curly braces like this to section the code. I think the readability is better this way.